pkg/telemetry/exporters/batch_processor.go

```diff
@@ -2,123 +2,191 @@ package exporters
 	return func(b *batchSpanProcessor) {
-		b.flushInternal = flush
+		if timeout > 0 {
+			b.timeout = timeout
+		}
 	}
 }
 
 type batchSpanProcessor struct {
-	bufferSize    uint32
-	flushInternal time.Duration
-
+	mt       sync.RWMutex
 	exporter trace.SpanExporter
-	batcher  batcher.Batcher
-	watcher  batcher.Watcher
+	maxSize  int
+	timeout  time.Duration
+	in       chan *trace.ReadOnlySpan
+	buffer   map[string][]trace.ReadOnlySpan
+	pointer  uuid.UUID
 }
 
-func NewBatchSpanProcessor(ctx context.Context, exporter trace.SpanExporter, opts ...BatchSpanProcessorOpt) (trace.SpanProcessor, error) {
-	processor := &batchSpanProcessor{
-		bufferSize:    defaultBatcherBufferSize,
-		flushInternal: defaultFlushInternal,
-		exporter:      exporter,
+func NewBatchSpanProcessor(ctx context.Context, exporter trace.SpanExporter, opts ...BatchSpanProcessorOpt) trace.SpanProcessor {
+	p := &batchSpanProcessor{
+		mt:       sync.RWMutex{},
+		exporter: exporter,
+		maxSize:  defaultBatchMaxSize,
+		timeout:  defaultBatchTimeout,
+		buffer:   map[string][]trace.ReadOnlySpan{},
+		pointer:  uuid.New(),
 	}
 
 	for _, apply := range opts {
-		apply(processor)
+		apply(p)
 	}
+	p.in = make(chan *trace.ReadOnlySpan, p.maxSize)
 
-	processor.batcher = batcher.NewBatcherWithBuffer(processor.bufferSize).WithFlushInterval(processor.flushInternal)
-	if err := processor.batcher.Start(ctx); err != nil {
-		return nil, fmt.Errorf("error starting batch processor: %w", err)
-	}
-	processor.watcher = batcher.NewWatcher(processor.onReady)
+	// start process loop
+	go p.run(ctx)
 
-	return processor, nil
+	return p
 }
 
 // No op
 func (b *batchSpanProcessor) OnStart(ctx context.Context, s trace.ReadWriteSpan) {}
 
 func (b *batchSpanProcessor) OnEnd(s trace.ReadOnlySpan) {
-	ctx := context.Background()
-
-	op := batcher.NewOperation(b.watcher, 1, s, true)
-
-	status := "success"
-	if err := b.batcher.Enqueue(op); err != nil {
-		// attempt to export directly if it's full or in the process of shutting down
-		// to ensure data is not lost
-		switch err {
-		case batcher.BufferFullError, batcher.BufferIsShutdown:
-			if err := b.exporter.ExportSpans(ctx, []trace.ReadOnlySpan{s}); err != nil {
-				status = "error"
-				logger.StdlibLogger(ctx).Error("error exporting span directly", "error", err)
-			}
-		default:
-			logger.StdlibLogger(ctx).Error("error enqueueing span for batch", "error", err)
-			status = "error"
-		}
-	}
-
-	metrics.IncrBatchProcessorEnqueuedCounter(ctx, metrics.CounterOpt{
-		PkgName: pkgName,
-		Tags:    map[string]any{"status": status},
-	})
+	// pass span into the channel
+	b.in <- &s
+	metrics.IncrBatchProcessorEnqueuedCounter(context.TODO(), metrics.CounterOpt{PkgName: pkgName})
 }
 
 func (b *batchSpanProcessor) Shutdown(ctx context.Context) error {
-	b.batcher.Flush()
+	if err := b.flush(ctx); err != nil {
+		logger.StdlibLogger(ctx).Error("error flushing spans on shutdown", "error", err)
+	}
```

Comment:
I'm not closing the `in` channel here, since executors can have long grace periods before they shutdown, and keeping the channel alive until it completely finishes helps with spans getting processed even if a `ctx.Done()` signal is received.

Would this be a problem is a channel is not closed before shutting down?

Reply:
So you're saying we need to keep `b.in` open to ensure that we don't close the channel before all traces are sent, right?

I think that sounds reasonable, unless we have a signal that notifies the batcher when the last job is fully processed. Otherwise, we just don't know the right moment to close the channel.

Reply:
yeah, services start shutting down after they get the termination signal in the container.
but it's unclear with the extended grace period, when exactly does `ctx.Done()` happens.

if it happens at the actual end when all the execution is done, then it won't be a problem. but if it happen any time before that, closing the channel prematurely will lead to lost spans.